### ISSUE TITLE
Accessibility Feedback - Design intent eBenefits - CH31 Orientation Content 

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#21478]
## Feedback
- **Must** = the feedback must be applied
- **Should** = the feedback is best practice
- **Consider** = suggestions/enhancements

https://staging.va.gov/careers-employment/vocational-rehabilitation/apply/introduction

### 1 .Content being added or removed from the page must announce a change to assistive devices [`defect-level-1`](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1) ❗️

Screen reader users will not know when content is being added or removed from the page unless it is explicitly announced. This is important for this page in particular since results are provided in dynamic grey boxes that are added and removed from the page.

- Use `aria-live` to announce new regions that screen reader users should be aware of while making choices in the wizard
- Fix bug related to orientation widget appearing but not disappearing upon certain wizard selections (to avoid screen reader users from accidentally using the orientation widget even when they don't need to)

### 2. Important unfocusable text in forms must be announced [`defect-level-1`](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1) ❗️

Most screen reader users will skip over text within forms that are not focusable e.g. `p`, `span`, `strong`. This means that important content provided within the grey boxes may be completely missed which can lead to unwanted behavior. In this particular context, it means that some screen reader users may completely skip over most sections of content within the orientation widget.

- Use `aria-describedby` to programmatically connect links within the grey boxes to their proceeding descriptions. For example, should a user tab onto "Apply online now with VA Form 28-1900," they should also hear, "Based on your answers, you probably qualify for VR&E benefits. Before you apply, please go through the VR&E orientation below. If you already know you want to apply for VR&E, you can go directly to the online application without going through the orientation below. "
-  For a proper fix, [I recommend using one thing per page](https://designnotes.blog.gov.uk/2015/07/03/one-thing-per-page/) so that results can be announced separately on their own page and the orientation can be completed on its own separate set of pages.

### 3. Visual groupings must be clear to screen reader users [`defect-level-1`](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1) ❗️

It is currently programmatically unclear when the wizard ends and the orientation wizard begins. Screen reader users may skip over the `h2` if they use the `tab` key to continue moving through the page as if it is a form.

- If no major changes can be made, a quick fix to this may be to add a `role="group"` to the orientation widget along with an `aria-labelledby` that points to the `h2` Veteran Readiness and Employment orientation. The `h2` "Veteran Readiness... orientation" should remain persistent across all of the orientation pages to act as a group label and create a logical heading order. The former `h2`s in each page (e.g. "VR&E basic benefit information" on the second page) should be demoted to `h3`s. Adding content that indicates progress and tying it to the `h3` using aria-describedby may help with understanding progression. This will make it more clear to screen reader users that each of these pages' information belong specifically to the orientation widget.
- [Reference ARIA17: Using grouping roles to identify related form controls](https://www.w3.org/TR/WCAG20-TECHS/ARIA17.html)

### 4. Navigation elements must be consistent within VA.gov [`defect-level-1`](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1) ❗️

Currently, the orientation widget uses the "back" and "continue" `button`s present in the form system. This may prompt screen reader users familiar with using VA forms online to expect that there are fields to be completed instead of just plain text and videos. This may result in them relying on the `tab` key to navigate, thereby missing all important information within the widget's pages.

In order to solve this, the team should use an alternative method of navigation that is not identical to the form system as it is not a form. For example "Continue Reading" instead of just "Continue" may make it more clear that fields are not being _saved_, and that they are expected to be parsing through text content.

### 5. The browser history MUST be updated if users expect to be able to use the browser "back" button [`defect-level-1`](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1) ❗️

Users may become frustrated if they lose their progress in the orientation widget should they use the browser back button to go back one step as opposed to the previous button in the widget. 

[Alternatively, making this more explicit by as a carousel or slideshow](https://www.w3.org/WAI/tutorials/carousels/) (e.g. using `aria-roledescription="carousel"`) along with `previous slide` and `next slide` buttons could also do the trick.

e.g.
<img width="591" alt="Screen Shot 2021-03-16 at 2 46 47 PM" src="https://user-images.githubusercontent.com/14154792/111363597-a18b9e00-8666-11eb-8d7e-87ef41583a60.png">

### Note on "must" recommendations

A11y specialists _strongly_ recommend that the VR&E orientation widget should be placed on its own single page _separate_ from the wizard itself as it would elegantly solve 3 out of the 4 launch blockers listed above. The recommendations we have provided earlier are to be noted as short term solutions in order to meet launch deadlines and basic compliance.

### Buttons should not be styled as links and links should not be styled as buttons [`defect-level-3`](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-3)

"Apply for Veteran Readiness and Employment" at the end of the VR&E orientation flow is a primary green button which is inconsistent with its link form in the grey box above the VR&E orientation.

<img width="759" alt="Screen Shot 2021-03-16 at 2 42 05 PM" src="https://user-images.githubusercontent.com/14154792/111363003-dfd48d80-8665-11eb-80ce-a5364e8f16b6.png">

## Consider
- In a future iteration, discuss with the stakeholders the possibility of moving their content around.
- Josh has comments below that should help describe some of these issues, we should reach out to him to review these, if needed.

## Acceptance Criteria
- Address "Must" items mentioned above, or note as platform requirement to complete:
  - [x] 1. Content being added or removed from the page must announce a change to assistive devices
    - [x] Request assistance from platform
  - [x] 2. Important unfocusable text in forms must be announced
    - [x] Request assistance from platform
  - [x] 3. Visual groupings must be clear to screen reader users
    - [x] Request assistance from platform
  - [x] 4. Navigation elements must be consistent within VA.gov
    - [x] Request assistance from platform
  - [ ] 5. The browser history MUST be updated if users expect to be able to use the browser "back" button
    - [ ] Request assistance from platform
- [x] A PR has been requested for merging

**Comment by joshkimux** [comment]
> For number 4, there is probably a multi-step approach: adjust the name of the button, but also look again at tab focus to be placed on the H2 after progression.

Correct! Adjust the names and actually set tab focus to the new `h3`s as the `h2` "VR&E Orientation" will be consistent throughout.

**Comment by joshkimux** [comment]
@jason-gcio , added another launch blocker which may be easier to describe over a call sometime. I'll need to do a little bit more research on this to see if there's an easier short term fix.

**Comment by joshkimux** [comment]
@jason-gcio 

Here's an example of how this might look like post the following changes:
- Add `role="group"` to the wizard and wrap orientation in a `section` with `aria-labelledby` pointing to their respective headers. This will help screen reader users understand that these are two separate sections on the page. View below video from :37 to :40.
- Use `aria-describedby` to announce important unfocusable text in the form. From :16 to :22 and from :24 to :28.
- Change copy of "Start VR&E Orientation" to "Start VR&E Orientation Slideshow" so it is more clear as to how this widget will operate.

https://user-images.githubusercontent.com/14154792/112011056-43dfd180-8afe-11eb-997e-670489f00183.mov

**Comment by micahchiang** [comment]
@joshkimux @jason-gcio - I've created a draft PR [here](https://github.com/department-of-veterans-affairs/vets-website/pull/16564) to incorporate as many of the a11y fixes suggested in this ticket as possible. Hoping to use that PR as a place to gain an extra pair of eyes from @joshkimux for feedback and collaboration as the work progresses.

**Comment by micahchiang** [comment]
Addressed point number 2 in [this commit](https://github.com/department-of-veterans-affairs/vets-website/pull/16564/commits/563d7c65736e903d398f0553f22549e50239485d)

**Comment by micahchiang** [comment]
Added step counter, updated button text, and added persistent H2 in [this commit](https://github.com/department-of-veterans-affairs/vets-website/pull/16564/commits/1e7075a45917d446c0be2c8e08de829ec7ccded5)

**Comment by micahchiang** [comment]
Went with the recommendation to add `aria-roledescription="carousel"` as opposed to fiddling with the browser back button. Effort there doesn't seem feasible right now, but possibly moreso if we end up exploring a separate page for the orientation at some point.

**Comment by joshkimux** [comment]
@micahchiang Thank you :D
Fingers crossed a separate page will happen in the near future as it's the most accessible option

**Comment by joshkimux** [comment]
Confirmed that Micah has successfully completed the acceptance criteria! Well done 🥳 🥇

**Comment by jason-gcio** [comment]
@noahgelman - per our conversation about #5, we are going to try a few solutions (page injection first, then maybe a new dedicated Drupal page, conversations with platform support, etc).  We recognize the importance of this and want to make sure we keep sight of it.  It was noted that we might be in production when the ultimate solution is implemented.  Thanks again for your help with all this!
